File: LinkedLists/ReverseList.java
package LinkedLists;

public class ReverseList {
  public ListNode reverseListIterative(ListNode head) {
    ListNode prev = null;
    while (head != null) {
      ListNode nextNode = head.next;
      head.next = prev;
      prev = head;
      head = nextNode;
    }

    return prev;
  }
  // Time complexity: O(n) we visit each node once
  // Space complexity: O(1)

  public ListNode reverseListRecursive(ListNode head) {
    if (head == null || head.next == null) {
      return head;
    }

    ListNode newHead = reverseListRecursive(head.next);
    head.next.next = head;
    head.next = null;
    return newHead;
  }

  // Time complexity: O(n) we visit each node once
  // Space complexity: O(n) due to recursive call stack
}

File: LinkedLists/reverse_list.py
from typing import Optional

from list_node import ListNode


class ReverseList:
    def reverseList(self, head: Optional[ListNode]) -> ListNode:
        prev = None
        while head:
            next_node = head.next
            head.next = prev
            prev = head
            head = next_node

        return prev
    
    def reverseListRecursive(self, head: Optional[ListNode]) -> ListNode:
        def reverse_list(prev: Optional[ListNode], current: Optional[ListNode]) -> ListNode:
            if not current:
                return prev
            
            next_node = current.next
            current.next = prev
            return reverse_list(current, next_node)
        
        return reverse_list(None, head)
    
    def reverseListRecursiveCanonical(self, head: Optional[ListNode]) -> ListNode:
        if not head or not head.next:
            return head
        
        new_head = self.reverseListRecursiveCanonical(head.next)
        head.next.next = head
        head.next = None
        return new_head
        
    # Time complexity: O(n) where n is the number of nodes in the linked list.
    # We visit each node exactly once to reverse the pointers.
    # Space complexity: O(1) for the iterative approach, 
    # as we only use a constant amount of extra space.
    # For the recursive approaches, 
    # the space complexity is O(n) due to the call stack.
